README.md
# hilbert-ui

A web interface to monitor, control and manage Dockapp. It includes a stand-alone web server 
that hosts a web service and a front end based on [React](https://facebook.github.io/react/). 

The parts of the ui are:

- Client
    - It's single page application based on the React framework.
    - Composed of static HTML, CSS, JS and image files.
    - It communicates with the server through AJAX queries.
    - ECMAScript 2015 is used and transpiled with babel.
    - Should be hosted with nginx.
- Server
    - Daemon that tracks the state of stations and operations
      performed on them.
    - Provides a web service interface (HTTP queries and JSON responses)
    - Programmed on node.js
    - Should be reverse proxied with nginx.

## Installation

0. Install [nginx](https://nginx.org/), if not already present in the system

0. Install node.js and yarn

0. Install the server's dependencies

        cd server
        yarn install --prod

0. Install the web front end's dependencies

        cd client
        yarn install --prod

0. Configure nginx to serve the static files and proxy the web service
 
    For example, within the `http` section of `nginx.conf` configure the default `server` with a 
    configuration like this one:
 
        server {
               listen       8080;
               server_name  localhost;
        
               location / {
                   root   /usr/local/hilbert/client/public;
                   index index.html index.htm;
               }
        
               location /api/ {
                       proxy_pass http://127.0.0.1:3000/;
                       proxy_redirect default;
                       proxy_http_version 1.1;
                       proxy_set_header Upgrade $http_upgrade;
                       proxy_set_header Connection 'upgrade';
                       proxy_set_header Host $host;
                       proxy_cache_bypass $http_upgrade;
               }
        
               error_page   500 502 503 504  /50x.html;
               location = /50x.html {
                   root   html;
               }
        }

    * Be sure to change the `root` to the location of the `public` directory of the hilbert ui. 
    * `proxy_pass` should correspond to the port the web service is running in (configured in the
      following step)
      
0. Copy server/config.json.sample to config.json and edit the file to match your system 
(see the Configuration section) below.
      
0. Start the web service by running

        node server/app/main.js

## Configuration

The ui configuration file is `server/config.json`. It supports the following options:

- `test`: (default: false) Set to `true` to run in test mode, which uses simulated stations.
- `port`: (default '3000') Port number the ui server will run in.
- `hilbert_cli`: Command line to run the hilbert-cli program (including the python interpreter).
- `hilbert_cfg`: Path to the Hilbert Cfg to use.
- `scriptConcurrency`: (default: 20) Number of simultaneous child processes that will be run.
- `max_log_length`: (default: 100) Max number of log entries to send through the web service.
- `log_directory`: (default: './log') Directory where logs will be saved.
- `log_level`: (default: 'info') Log level. Set to one of `error`, `warn`, `info`, `verbose`, `debug` or `silly`.
- `mkls_poll_delay`: (default: 1000) Number of milliseconds to wait between polls to MKLivestatus.
- `mkls_cmd`: (default: 'nc localhost 6557') Command line used to connect to MKLivestatus.
- `operation_timeout`: (default: 600) Number of seconds to wait for station operations (stop, start, change app) to finish before timing out.
- `error_lock_time`: (default: 30) Number of seconds a station will be locked after an operation fails.
- `db_path`: (default: '') Path to the sqlite database file for persistant storage of features such as presets. 

## Presets

The client provides a **Preset** feature that allows to quickly change application assignments for all the
stations at once. For instance you can save a "School visit" preset to prepare the stations for such
an event (and then restore them to default after), or a "Development" preset for testing out software for
a new exhibition. A **Preset** stores the current assignment of applications with a descriptive name.

To create a **Preset** first start the desired applications in each station. Once done select "Create Preset" 
in the **Preset** dropdown menu in the top bar. The system will ask for a name and save your preset.

You can activate a **Preset** by choosing it in the Preset selector and then selecting "Activate"
in the dropdown menu. The system will automatically switch the applications in all stations to match the
saved assignment.

Note that Presets only include application assignments and not the On/Off state of stations. Activating
a preset will never start nor stop stations. Only stations that were On and had an application running 
are saved in the preset. Also, when activating a preset it will only affect the stations that are part of 
it; i.e. it won't have any effect on stations that were Off when the preset was created or that were added 
to Hilbert later.

You can modify an existing Preset by choosing it in the Preset selector and then selecting "Save" in the
dropdown menu. The current application assignment will be saved under that name, just like creating a new
preset but without being prompted for a name.

Finally you can delete a preset by choosing it in the selector and then selecting "Delete" in the dropdown
menu.

## Directory reference

- **client**: Client files
    - **src/**: Client sources (JSX / ECMAScript 2015)
    - **sass/**: SASS stylesheets, compiled through `gulp sass`
    - **public/**: Static web files for the client, including CSS, JS, images and 
          third party libraries
- **server/**: Server files
    - **api-docs**: API documentation
    - **app/**: Transcompiled JS files. These are generated by babel and shouldn't be edited directly.
    - **app/test**: Transcompiled test files. These are generated by babel and shouldn't be edited directly.
    - **log/**: Back end log files
    - **scripts/**: Auxiliary scripts to interact with hilbert-cli
    - **src/**: Server sources (ECMAScript 2015)
    - **src/test**: Test sources (ECMAScript 2015)
    - **data/**: Auxiliary data files
    - **config.json**: Back end configuration file

## Building the UI client

The client sources are compiled from JSX / ECMAScript 2015 to regular JS using Babel. 
Dependencies (like React) are also compiled and packed to a vendors.js file. 

0. Install the dependencies

        cd client
        yarn

0. Compile JS sources (from the `client` directory)

        gulp scripts

0. Compile SASS stylesheets (from the `client` directory)

        gulp styles

## Deployment with Docker

The [Dockerfile](Dockerfile) can be used to build a self-contained Docker image for the ui. 
Install [Docker](https://www.docker.com/), then build the image from the top level directory:

    docker build -t hilbert-ui .

Spin up a Docker container with

    docker run -p 8080:8080 hilbert-ui

This exposes the container's port 8080 to the host's port 8080.

## API

See `api-docs/openapi.yaml` for the full UI Server API documentation in Swagger / OpenAPI v2 format. 
It can be read via http://editor.swagger.io/ 

## Credits

Eric Londaits for [IMAGINARY](https://www.imaginary.org)

## Copyright and license

Code and documentation copyright 2016 IMAGINARY gGmbH 
Code and documentation released under the Apache [License](LICENSE.md).
See the [Notice](NOTICE.md) for license information of included 
dependencies.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/) 
and this project adheres to [Semantic Versioning](http://semver.org/).

## [v0.22.0] 2019-07-20
## Added
- Adds station name filter in the dashboard.
- Adds support for opening services with URLs in the quick action menu.
- Adds logo and favicon.
- Adds new application icons.
- Added /services endpoint to the HTTP API.
## Fixed
- Fixes bug where presets failed to create / update if stations were off.

## [v0.21.1] 2019-07-04
## Fixed
- Fixes critical bug querying MK Livestatus

## [v0.21] 2019-07-03
## Added
- Adds a "Quick menu" to stations in the UI Client.
- Adds "Restart" and "Restar app" commands to stations.
- Prints client version in the console on start.

## [v0.20.1] 2019-06-27
## Changed
- Downgrades node engine requirement to v10.14

## [v0.20] 2019-06-24
## Changed
- Updates node to v10.16.0 and many dependencies. 
- Removes need for server compilation. Updates client compilation.
## Fixed
- Fixes (erroneous) function calls that failed after updating Set polyfill.- Fixes Client regression bug: "No connection to server" message doesn't disappear when connection comes back up.


## [v0.19.1] 2018-10-14
## Fixed
- Makes hilbert-cli exec buffers larger to cope with large cfgs

## [v0.19.0] 2018-04-23
### Added
- Responsive support
### Fixed
- Various UI adjustments and bugfixes

## [v0.18.0] 2018-04-05
### Added
- Shows timestamp of last CheckMK status read in the /server/mklivestatus api
### Fixed
- More resilient parsing of app state in Check MK

## [v0.17.0] 2018-04-03
### Added
- Adds a new layout. It's so different it's unfair to put it in a "Changed" section.
- Adds a new menu for reordering stations by various fields
- Adds a menu option for changing the app of the selected stations
- Adds a dropdown menu for changing app in each station

## [v0.16.0] 2018-04-03
### Added
- Adds "info" button to stations which shows station ID and description

## [v0.15.1] 2018-04-02
### Fixed
- Fixes bug that prevented receiving state from CheckMK

## [v0.15] 2018-04-02
### Added
- Ignore stale app data from CheckMK

## [v0.14] 2018-04-02
### Added
- Adds a System menu to start and stop stations
- Adds a Test menu to simulate errors in test mode
- Adds notifications to the client
- Stations now temporarily lock after certain errors
- Station operations now time out after a configurable amount of time
- Adds the 'operation_timeout' and 'error_lock_time' config keys.
### Changed
- The server now properly handles a number of error conditions, specially during operations
- The client "Event log" is now called "Latest notifications"

## [v0.13.1] 2018-03-04
### Added
- Adds schema validation to the HTTP API

## [v0.13] 2018-02-20
### Added
- Added support for the new Hilbert CLI configuration format
- Added /applications and /station_profiles entry points to the API
- Shows application and station names instead of IDs in the the client
- Uses station profiles instead of type for filtering in the client
- Added API documentation in Swagger / OpenAPI format
### Changed
- Call Hilbert CLI directly instead of through wrapper scripts


## [v0.12.1] 2018-01-30
### Fixed
- Presets are deselected after deleting them

## [v0.12] 2018-01-30
### Added
- Added a header bar to the UI
### Changed
- Presets were changed to always map all stations instead of subsets

## [v0.11.1] 2018-01-28
### Fixed
- Removes internal properties from the station data sent through the HTTP API

## [v0.11] 2018-01-28
### Added
- Added support for Presets (named station->app mappings)
- Added new API for modules that add functionality to the server

## [v0.10] 2016-12-06
### Added
- CLI program (hilbert-http-api) that can list, start and stop stations.
- Added tests for the HTTP API layer
- Added interface to the Test Backend for defining test data programmatically
- Added LICENSE and NOTICE
- Added Travis CI configuration
### Changed
- Makes delays in test mode optional
### Fixed
- HTTP Api returns error instead of throwing exceptions when required
  arguments are missing.

## [v0.9] 2016-09-30
### Added
- Documented HTTP API
- Test entry point with dummy test
### Changed
- New HTTP API in server

## [v0.8] 2016-09-24
### Changed
- The project was renamed to hilbert-ui.
- The client files were moved to the `client` directory.
- Updated documentation.

## [v0.7.1] 2016-09-17
### Added
- This changelog was added. Some version numbers don't perfectly adhere
  to semantic versioning (arguably) but things will improve from now on
  (hopefully).
- Added undocumented configuration keys to README.md

## [v0.7] 2016-08-30
### Added
- Adds a "Terminal log" button to each station in the UI and removes
  the previous external button. 

## [v0.6] 2016-08-28
### Added
- The UI shows the reason why a station is down (manual shutdown or 
unexpected)

## [v0.5.1] 2016-08-28
### Changed
- The internal STARTING station state is split into STARTING_STATION and
  STARTING_APP.
### Fixed
- Fixes bad handling of return codes in wrapping scripts.

## [v0.5] 2016-08-27
### Changed
- Interfaces to Dockapp through high level wrapping scripts.
- Changes to the output of 'debug' and 'verbose' level logs. 
### Fixed
- Fixes errors in parsing MKLivestatus responses.

## [v0.4.2] 2016-08-25
### Fixed
- Correctly parse station configurations that include comments.

## [v0.4.1] 2016-08-25
### Changed
- The console output respects the configured logging level. 

## [v0.4] 2016-08-21
### Added
- The output of scripts run on the server host can be viewed in a 
terminal window.
- Added various monitoring methods to the http API.

## [v0.3] 2016-08-15
### Added
- The output of scripts run on each terminal can be viewed in a terminal
window.
- A prominente message bar in the UI informs when the connection to the
server is lost.
- The client increases the poll time when it fails to contact the 
server.

server/lib/main.js
const logger = require('winston');
const nconf = require('nconf');
const appPackage = require('../package.json');
const StationManager = require('./lib/station-manager');
const HilbertCLIConnector = require('./lib/hilbert-cli-connector');
const MKLivestatusConnector = require('./lib/mk-livestatus-connector');
const HttpAPIServer = require('./lib/http-api-server');
const TestBackend = require('./lib/test-backend/test-backend');
const testHilbertCfg = require('../data/test_mode/cfg.json');

nconf.env().argv();
nconf.file('config.json');
nconf.defaults({
  port: '3000',
  test: false,
  scriptConcurrency: 20,
  max_notifications: 100,
  log_directory: './log',
  log_level: 'info', // error, warn, info, verbose, debug, silly
  mkls_poll_delay: 1000,
  mkls_cmd: 'nc localhost 6557',
  long_poll_timeout: 15,
  operation_timeout: 600,
  error_lock_time: 30,
  db_path: '',
});

logger.add(logger.transports.File, {
  filename: `${nconf.get('log_directory')}/hilbert-ui.log`,
  level: nconf.get('log_level'),
  handleExceptions: true,
  json: false,
});

logger.level = nconf.get('log_level');

process.on('uncaughtException', (err) => {
  logger.error(`Uncaught exception: ${err.message}. Exiting process.`);
  process.exitCode = 1;
});

logger.info(`Starting hilbert-ui server (v${appPackage.version})`);

let hilbertCLIConnector = null;
let mkLivestatusConnector = null;

if (nconf.get('test')) {
  logger.info('Running in Test Mode');
  const testBackend = new TestBackend(nconf, logger);
  testBackend.simulateDelays = true;
  testBackend.load(testHilbertCfg);
  hilbertCLIConnector = testBackend.getHilbertCLIConnector();
  mkLivestatusConnector = testBackend.getMKLivestatusConnector();
} else {
  hilbertCLIConnector = new HilbertCLIConnector(nconf, logger);
  mkLivestatusConnector = new MKLivestatusConnector(nconf, logger);
}

const stationManager = new StationManager(
  nconf,
  logger,
  hilbertCLIConnector,
  mkLivestatusConnector
);

let server = null;
stationManager.init()
  .then(() => {
    server = new HttpAPIServer(stationManager, nconf, logger);
    return server.init();
  })
  .then(() => {
    server.listen(nconf.get('port'));
  })
  .catch((err) => {
    logger.error(`Error initializing Station Manager: ${err.message}. Exiting process.`);
    process.exitCode = 2;
  });
